APEX-92 #Comment #resolve

ishark opened the pull request:
Fix for adding failed nodes to blacklist, after failure is observed for the same node more than MAX number of tries

243826 commented:
David, isn't there a way to check if the app has ended?

davidyan74 commented:
There is.  I can add a check and use different logging level depending on whether the app has ended.  However, I think this would be an overkill.

243826 commented:
It's necessary. If platform masks the problems then it is more likely to
become platform problem in users eyes. If the normal environment is not
encountered, the platform should be loud about it.

On Tue, Jul 21, 2015 at 10:52 AM, David Yan <notifications@github.com>
wrote:

> There is. I can add a check and use different logging level depending on
> whether the app has ended. However, I think this would be an overkill.
>
> —
> Reply to this email directly or view it on GitHub.
>

Commit:
APEX-92 #Comment #resolve
Fix for adding failed nodes to blacklist, after failure is observed for the same node more than MAX number of tries

gauravgopi123 commented on engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java:
```diff
@@ -910,6 +927,9 @@ private void execute() throws YarnException, IOException
+      if(!blacklistAdditions.isEmpty()) {
+        amRmClient.updateBlacklist(blacklistAdditions, null);
+      }
       if (dnmgr.forcedShutdown) {
```
Should blacklisted hosts not be periodically checked to remove from blacklist?

vrozov commented on engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java:
```diff
@@ -880,6 +883,20 @@ private void execute() throws YarnException, IOException
             numFailedContainers.incrementAndGet();
+            if (exitStatus != 1) {
+              // If container failure due to framework
+              AtomicInteger failedTimes = new AtomicInteger(1);
```
Is it necessary to use AtomicInteger here? What is the other thread that uses the same integer?

tweise replied:
No, it is not necessary.

vrozov commented on engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java:
```diff
@@ -880,6 +883,20 @@ private void execute() throws YarnException, IOException
+              // If container failure due to framework
+              AtomicInteger failedTimes = new AtomicInteger(1);
+              String hostname = allocatedContainer.container.getNodeId().getHost();
+              if (failedContainersMap.containsKey(hostname)) {
```
Using ConcurrentMap does not prevent thread switch here. In case failedContainersMap does not have mapping for hostname, there is race condition here. You may use putIfAbsent() if the map can be accessed from two or more different threads.

tweise replied:
No need for ConcurrenMap.

ishark replied:
ConcurrentMap is needed only if later down the line, we want to expose failure count per hostname as part of ClusterAppStats. Similar to getFailedContainers (returns total number of failures)

gauravgopi123 commented on engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java:
```diff
@@ -880,6 +883,20 @@ private void execute() throws YarnException, IOException
+              String hostname = allocatedContainer.container.getNodeId().getHost();
+              if (failedContainersMap.containsKey(hostname)) {
+                failedTimes = failedContainersMap.get(hostname);
+                int failureOnNode = failedTimes.incrementAndGet();
```
Should the hostname not be removed from failedContainerMap once the container is allocated and works fine.. I mean what happens if during first 4 tries the container fails but in 5th try container works on host X.. I dont' see it getting removed from the failedContainerMap

vrozov replied:
I'd suggest instead of removing hostname, reset count to 0. Adding and removing entries in a hash map are somewhat costly.

gauravgopi123 replied:
Since the map will not be too big.. I think removing should be fine..

vrozov replied:
If later we decide to track not only counter of failed attempts and add for example time of the last failed attempt or keep the full list of failed attempts, keeping the entry is preferable. It also leads to less garbage collection and because the map will not too big it should be OK to keep the entry and reset the counter to 0.

tweise replied:
Garbage collection is not a concern here. Container allocation is an infrequent operation.

gauravgopi123 commented on engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java:
```diff
@@ -100,6 +100,7 @@
   private static final long DELEGATION_TOKEN_RENEW_INTERVAL = Long.MAX_VALUE / 2;
   private static final long DELEGATION_TOKEN_REMOVER_SCAN_INTERVAL = 24 * 60 * 60 * 1000;
   private static final int NUMBER_MISSED_HEARTBEATS = 30;
+  private static final int MAX_CONTAINER_FAILURES_PER_NODE = 10;
```
Should this not be taken as an property from the user? Default can be 10

vrozov replied:
10 as default seems to be high. If 3 consecutive attempts fail, most likely something wrong with the node.

gauravgopi123 replied:
Even I was thinking same that 10 is too high and user should have some control over it.

vrozov replied:
Agree, can the default be changed to 3 and provide ability for end user to overwrite the default value.

vrozov commented on engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java:
```diff
@@ -880,6 +883,20 @@ private void execute() throws YarnException, IOException
+                int failureOnNode = failedTimes.incrementAndGet();
+                if (failureOnNode > MAX_CONTAINER_FAILURES_PER_NODE) {
+                  // Blacklist the node
+                  blacklistAdditions.add(hostname);
```
Should this be logged or the event is logged somewhere else?

otterc commented on engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java:
```diff
@@ -20,9 +20,10 @@
 import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.concurrent.ConcurrentMap;
-import java.util.concurrent.atomic.AtomicInteger;
+import java.util.Map.Entry;
```
This is throwing error because in ASCII caps M comes before lowercase 'c' :)

ishark replied:
You are right :) But if I move it before java.util.concurrent the checkstyle errors are increased..

vrozov commented on engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java:
```diff
@@ -880,6 +898,19 @@ private void execute() throws YarnException, IOException
+            if (exitStatus != 1) {
+              // If container failure due to framework
+              String hostname = allocatedContainer.container.getNodeId().getHost();
+              int failedTimes = failedContainersMap.putIfAbsent(hostname, 1);
```
Subject to NullPointerException due to implicit unboxing.

vrozov commented on engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java:
```diff
@@ -880,6 +898,19 @@ private void execute() throws YarnException, IOException
+              String hostname = allocatedContainer.container.getNodeId().getHost();
+              int failedTimes = failedContainersMap.putIfAbsent(hostname, 1);
+              if (failedTimes > 1) {
+                failedContainersMap.replace(hostname, ++failedTimes);
```
Generally speaking not thread safe if this function will be invoked from two different threads with containers pointing to the same hostname.

tweise replied:
This is not intended to to be invoked from multiple threads.

tweise commented on engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java:
```diff
@@ -651,6 +647,9 @@ private void execute() throws YarnException, IOException
     int minVcores = conf.getInt("yarn.scheduler.minimum-allocation-vcores", 0);
     LOG.info("Max mem {}m, Min mem {}m, Max vcores {} and Min vcores {} capabililty of resources in this cluster ", maxMem, minMem, maxVcores, minVcores);
 
+    int maxConsecutiveContainerFailures = conf.getInt("MAX_CONSECUTIVE_CONTAINER_FAILURES", MAX_CONTAINER_FAILURES_PER_NODE);
```
This looks like a hack. How do these configuration values make it into the conf object?

gauravgopi123 replied:
This doesn't work
